### DISCUSSION TITLE
questions about COPY FROM

### DISCUSSION THREAD

**Discussion by jerch** [psycopg/psycopg#677] [General]
I am in the process to port the `django-fast-update` to psycopg3, esp. this code here https://github.com/netzkolchose/django-fast-update/blob/master/fast_update/copy.py

The psycopg2 code does the value formatting on python side for the TEXT format. Following the hints regarding a possible psycopg3 rewrite with `write_row` @dvarrazzo gave me [here](https://github.com/psycopg/psycopg2/discussions/1438#discussioncomment-2414208), psycopg3 does an almost perfect job in automatic value formatting out-of-the-box and the update task runs almost 2-3x faster with the C-binding than my old python-side formatting, nice!

Now the questions:
- Is `FORMAT BINARY` worth a shot perf/bandwidth-wise? Is the binary protocol interface stable enough across postgres versions to make it the default? (Asking prehand, as I'd have to implement more strict field adoption in that case...)
- django currently does not set the dumper for `jsonb` to `Jsonb` - is there a way to set a dumper only on the active cursor object without interfering with module globals? Or do I have to decorate all json value with `Jsonb(value)` manually?

Last but not least - thx for the much better COPY support in psycopg3 and the speed enhancements!

**Comment by dvarrazzo** [comment]
Hello! I'm glad the porting is promising.

> - Is FORMAT BINARY worth a shot perf/bandwidth-wise? Is the binary protocol interface stable enough across postgres versions to make it the default? (Asking prehand, as I'd have to implement more strict field adoption in that case...)

I am not sure about using the binary format for generic work such as Django, because you need to have a real 1:1 mapping between postgres fields. I don't know your project but, if it makes use of the django fields definitions, there might be discrepancies between them and the postgres definitions.

Maybe you can give it a shot and see what changes in terms of performances and what errors you stumble upon during testing, and see if it's an acceptable trade-off for you.

Answering just in regards of the data stability, I don't know what type of backward compatible guarantee the Postgres people give to the binary format, but I haven't seen any difference in the 9.6-16 postgres range and I doubt there were before that. The only foreseeable problem type-wise, e.g. an int64 postgres field will not accept an int32 postgres value in binary copy and will throw an error because it was expecting 8 bytes instead of 4.

> is there a way to set a dumper only on the active cursor object without interfering with module globals?

Yes, you can register a dumper on a cursor only, and it will be used only for the copy operations run on that cursor. Some documentation [here](https://www.psycopg.org/psycopg3/docs/advanced/adapt.html#dumpers-and-loaders-life-cycle).

**Reply by jerch** [reply]
> I am not sure about using the binary format for generic work such as Django...

Yeah still trying to find out, if Django exhibts enough details. For int types I can defer the needed byte type from `db_type()`, which would give me `smallint` | `integer` | `biginteger`. Will see how that works out for more complicated types like numeric, arrays or hstore.

> Answering just in regards of the data stability, I don't know what type of backward compatible guarantee the Postgres people give to the binary format, but I haven't seen any difference in the 9.6-16 postgres range and I doubt there were before that.

Ah ok - sounds to me like it is worth to give it a try, yeah. Was just asking because the postgres docs contains a warning regarding cross version compat.

Thx again!

(Edit - Did github remove the "mark as answer" button? Cant find it, well your reply is the answer...)

**Reply by jerch** [reply]
Small update: Binary is ~ 5-10% faster for primitive types in my tests. The tests are not yet adjusted to different TEXT width in int fields from big numbers, thus my expectation is that the perf bonus should get bigger for more random value distribution. Also the more complicated fields are not handled yet. Still with this first glance binary looks promising...

**Comment by jerch** [comment]
@dvarrazzo Sorry for bugging you again.

I've made good progress with copy & BINARY, it gives a reasonable speedup in my local tests to try to use it by default. I still have a few questions in regards to BINARY:
- I cannot find docs about the binary format, so I'm not quite sure, which type names will work in `copy.set_types(list_of_types)`. Where can I get a list of supported type names from? Should I look at the keys of registered dumpers?
- Django gives me with `db_type()` things like `varchar(255)` which I was able to get working in `set_types` with either `varchar` or `text` by testing around. Here I wonder if there is an official "cross-cast" compat list somewhere (it is not obvious to me, why `text` is compatible to `varchar(xy)` in the binary protocol). I also saw a tiny perf advantage of 5-10% by using `text` - is just a test artifact or justified by a different control flow in psycopg, or known to be faster on postgres side?
- How would psycopg handle self-defined composite types with BINARY? I saw that psycopg refers internally to OIDs, thus sees all registered types of a postgres instance. Does psycopg provide enough primitives to build own binary dumpers for self-defined composite types as well? Or are there edge cases, that cannot be expressed in binary? (I've never seen anyone doing composite types with custom django fields, so this question is more about possible limitations for very edgy implementations and whether I should fallback to TEXT there).

Edit: A slightly different question regarding COPY and `set_types` - I cannot get 'jsonb' working w'o decorating all values as `Jsonb(value)`. Is there a more straight forward method to force copy into using a certain dumper solely from the column type w'o that additional step? In theory I know exactly the expected column types at `copy.write_row()`, thus I dont quite understand, why the additional `Jsonb()` step is needed (Sidenote - for Int2|Int4|Int8 it seems it is not needed?)

**Reply by dvarrazzo** [reply]
Hello @jerch, let's see if I can answer...

> - I cannot find docs about the binary format, so I'm not quite sure, which type names will work in `copy.set_types(list_of_types)`. Where can I get a list of supported type names from? Should I look at the keys of registered dumpers?

The types are in the [types registry](https://www.psycopg.org/psycopg3/docs/api/types.html#psycopg.types.TypesRegistry) attached to the context you have customized (global -> connection -> cursor). Types are just informative, there aren't necessarily adapters attached to them.

```python
>>> psycopg.adapters.types["text"]
<TypeInfo: text (oid: 25, array oid: 1009)>
```

The adapters are in the the dumpers and loaders of the `AdaptersMap` of the context; you can check which ones are defined after you have the oid of the type:

```python
>>> psycopg.adapters.get_dumper_by_oid(psycopg.adapters.types["text"].oid, psycopg.pq.Format.BINARY)
psycopg_c._psycopg.StrBinaryDumper

>>> psycopg.adapters.get_loader(psycopg.adapters.types["text"].oid, psycopg.pq.Format.BINARY)
psycopg_c._psycopg.TextBinaryLoader
```


> - Django gives me with `db_type()` things like `varchar(255)` which I was able to get working in `set_types` with either `varchar` or `text` by testing around. Here I wonder if there is an official "cross-cast" compat list somewhere (it is not obvious to me, why `text` is compatible to `varchar` in the binary protocol). I also saw a tiny perf advantage of 5-10% by using `text` - is just a test artifact or justified by a different control flow in psycopg, or known to be faster on postgres side?

What you call "cross-cast" is what the Postgres backend refers to as "binary coercible". There is info in the [`pg_cast`](https://www.postgresql.org/docs/current/catalog-pg-cast.html
) system table. Here are a few:

```
piro=# select castsource::regtype, casttarget::regtype from pg_cast
       where castmethod = 'b' and castsource = any('{text,varchar}'::regtype[]);
┌───────────────────┬───────────────────┐
│    castsource     │    casttarget     │
├───────────────────┼───────────────────┤
│ text              │ character         │
│ text              │ character varying │
│ character varying │ text              │
│ character varying │ character         │
└───────────────────┴───────────────────┘
```

the information is not explicitly reported in the psycopg types info. Querying the registry to figure out which binary coercible types are supported:

```python
In [24]: bcasts = conn.cursor(row_factory=namedtuple_row).execute("""
    ...: select castsource as srcoid, casttarget as tgtoid,
    ...: castsource::regtype::text as srcname, casttarget::regtype::name as tgtname
    ...: from pg_cast where castmethod = 'b'
    ...: order by 3, 4""").fetchall()

In [25]: for cast  in bcasts:
    ...:     try:
    ...:         tsrc = psycopg.adapters.types[cast.srcoid]
    ...:         ttgt = psycopg.adapters.types[cast.tgtoid]
    ...:         dsrc = psycopg.adapters.get_dumper_by_oid(cast.srcoid, psycopg.pq.Format.BINARY)
    ...:         ltgt = psycopg.adapters.get_loader(cast.tgtoid, psycopg.pq.Format.BINARY)
    ...:         assert ltgt, f"no loader for {cast.tgtname}"
    ...:     except Exception as ex:
    ...:         pass
    ...:         # Uncomment to see why the pair is not supported
    ...:         # print(" " * 20, src, tgt, type(ex).__name__, str(ex))
    ...:     else:
    ...:         print(f"{cast.srcname} ({dsrc.__name__}) => {cast.tgtname} ({ltgt.__name__})")
    ...: 
character varying (StrBinaryDumperVarchar) => character (TextBinaryLoader)
character varying (StrBinaryDumperVarchar) => text (TextBinaryLoader)
cidr (NetworkBinaryDumper) => inet (InetBinaryLoader)
integer (Int4BinaryDumper) => oid (OidBinaryLoader)
oid (OidBinaryDumper) => integer (Int4BinaryLoader)
text (StrBinaryDumper) => character (TextBinaryLoader)
text (StrBinaryDumper) => character varying (TextBinaryLoader)
```

Maybe the difference of speed you are seeing is caused by the need for boundary checks for the varchar type and not for the text type. The `pg_type` catalog table maps the types to the load function:
```
piro=# select oid::regtype, typreceive from pg_type where oid = any('{text,varchar}'::regtype[]);
┌───────────────────┬─────────────┐
│        oid        │ typreceive  │
├───────────────────┼─────────────┤
│ text              │ textrecv    │
│ character varying │ varcharrecv │
└───────────────────┴─────────────┘
```

these are C functions in the backend code (in `src/backend/utils/adt`): probably varchar requires some access to the registry to handle the type modifier which you found measurable in your benchmark.

> - How would psycopg handle self-defined composite types with BINARY? I saw that psycopg refers internally to OIDs, thus sees all registered types of a postgres instance. Does psycopg provide enough primitives to build own binary dumpers for self-defined composite types as well? Or are there edge cases, that cannot be expressed in binary? (I've never seen anyone doing composite types with custom django fields, so this question is more about possible limitations for very edgy implementations and whether I should fallback to TEXT there).

The composite adapters use the types registry of the context where they are used. If the `cursor.adapters.types` know the type, and `cursor.adapters` have binary dumper/loader for the type, then the composite should be handled correctly.

Hope this helps!

**Reply by jerch** [reply]
Thx alot, this gives me enough details to lift all my uncertainty edges.
